### DISCUSSION TITLE
Non-Visual / Picture Data

### DISCUSSION THREAD

**Discussion by GGDRriedel** [ultralytics/ultralytics#19945] [Ideas]
Hi there. 
We want to use the system to handle crops of very large scale geophysical data. 

For that, we would be interested in the system handling direct data input. 

As for now, I am forced to generate pictures from the data, input that for training and feeding detection some pseudo RGB-arrays. 

Basically I am wrapping the detection in an IO system. 

Is there any way I can skip that? For example providing data handlers for learning input would already make things much easier than having to generate hundreds of thousands of PNGs. 

I would be thankful for some input.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @GGDRriedel, thank you for reaching out and sharing your use case with us 🚀! 

It sounds like you're working on an exciting project involving geophysical data. While YOLO is primarily designed to work with image-based inputs, there might be ways to streamline your workflow. An Ultralytics engineer will review your inquiry soon to provide more detailed guidance.

In the meantime:
- If this is a feature request or enhancement, feel free to provide additional details about your specific dataset and requirements to help us better understand your needs.
- If you're encountering any issues or bugs, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to clarify the problem and help us debug efficiently.

We also recommend exploring the [Ultralytics Docs](https://docs.ultralytics.com/) for insights into supported data formats and potential customizations. You may find relevant information in the [Python](https://docs.ultralytics.com/usage/python/) or [CLI](https://docs.ultralytics.com/usage/cli/) usage sections.

### Upgrade

Ensure you're using the latest version of the `ultralytics` package with all dependencies updated. You can upgrade in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) as follows:

```bash
pip install -U ultralytics
```

### Community and Resources

Join the Ultralytics community to discuss ideas or ask questions:
- [Discord](https://discord.com/invite/ultralytics) for real-time conversations 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to collaborate with other users

### Verified Environments

If you're experimenting with different setups, you can run YOLO in these verified environments (with preinstalled dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn)):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Stay tuned, and we’ll get back to you with further insights soon! 😊

**Comment by glenn-jocher** [comment]
@GGDRriedel thanks for your question! For non-image data, you can use in-memory numpy arrays or PyTorch tensors directly with YOLO models by passing them to the `predict()` method. The Ultralytics [supported formats documentation](https://docs.ultralytics.com/usage/simple-utilities/) shows how to handle various input types programmatically. This avoids needing to save intermediate files while maintaining compatibility with our data pipelines.

**Reply by GGDRriedel** [reply]
The supported formats section on the page you linked shows picture and video formats. 

I'd much rather use a command line telling it to just use some custom dataloaders I define, sort of like a shader model in graphics, than diving through your code to find the dataloader cores that I need to load and implement for training. After all, I am not a programmer. 

Most of the time (at least in Geophysics) we have "giant" files (several gigabytes) of data that gets sliced into profiles and a few kb of "anotation" like first break picks or markings in the data. 
Here is an example for seismics: 

https://github.com/DaloroAT/first_breaks_picking

For me it's patterns in GPR aka boxes.

**Reply by glenn-jocher** [reply]
@GGDRriedel you can create custom data handlers by extending our BaseDataset class. While we don't currently support CLI configuration of custom loaders, you can implement your geophysical data loading logic in Python by following our [custom dataset documentation](https://docs.ultralytics.com/datasets/detect). This allows direct loading of array data without intermediate PNG conversion while maintaining compatibility with YOLO's training pipeline.
